### PR TITLE
Don't allow users to join group without accepting its terms.

### PR DESCRIPTION
**Pull request by johngian**

### CODE REVIEW

**Comment by johngian** [comment]
@akatsoulas r?

**Comment by akatsoulas** [comment]
Reminder: if PR #1250 is merged first, then the migration here needs update and vice versa.

File: mozillians/groups/urls.py [hunk 1]
```diff
@@ -12,6 +12,12 @@
     url('^groups/add/$', 'views.group_add_edit', name='group_add'),
     url('^group/(?P<url>[-\w]+)/edit/$', 'views.group_add_edit', name='group_edit'),
     url('^group/(?P<url>[-\w]+)/delete/$', 'views.group_delete', name='group_delete'),
+    url('^group/(?P<url>[-\w]+)/terms/$', 'views.group_terms',
+        name='group_terms'),
+    url('^group/(?P<url>[-\w]+)/terms/accept$', 'views.accept_group_terms',
+        name='accept_group_terms'),
+    url('^group/(?P<url>[-\w]+)/terms/deny$', 'views.deny_group_terms',
+        name='deny_group_terms'),
```
Comment (akatsoulas) [inline mozillians/groups/urls.py]: Can't you merge accept and deny in a single view?

Comment (johngian) [inline mozillians/groups/urls.py]: I don't see why it's better to check ``if foo == 'accept' elif foo == 'deny'``. This way is much cleaner.

**Comment by johngian** [comment]
@akatsoulas PR updated after #1268

File: mozillians/groups/urls.py (hunk 1 above)
Comment (akatsoulas) [inline mozillians/groups/urls.py]: I would suggest to use a single url with a single view. Less fragmentation, thus easier to follow what the code is doing.

File: mozillians/groups/views.py [hunk 2]
```diff
@@ -265,6 +272,42 @@ def edit(request, url, alias_model, template):
...
+def group_terms(request, url):
+    """Group terms page."""
+    group = get_object_or_404(Group, url=url)
+    if not group.terms:
+        return redirect(reverse('groups:show_group', args=[group.url]))
+    return render(request, 'groups/terms.html', {'group': group})
+
+
+def accept_group_terms(request, url):
+    """Accept group terms."""
```
Comment (akatsoulas) [inline mozillians/groups/views.py]: Please be a bit more descriptive on what the view is doing.

**Comment by akatsoulas** [comment]
Probably you need to update the `has_member` method in the `groups/models.py` to include the `has_terms` flag.

**Comment by akatsoulas** [comment]
What happens regarding the acceptance of terms, if exist, for the creators of a group?

File: mozillians/groups/models.py [hunk 3]
```diff
@@ -196,6 +197,7 @@ class Group(GroupBase):
                    u'curator a reminder')
     )
 
+    terms = models.TextField(default='', verbose_name=_('Terms'))
```
Comment (akatsoulas) [inline mozillians/groups/models.py]: `blank=True`?

Comment (akatsoulas) [inline mozillians/groups/models.py]: Also this is missing from the form in edit groups.

**Comment by johngian** [comment]
PR updated
* Group creator accepts terms directly when creating a group
* I added some more tests

File: mozillians/templates/groups/terms.html [hunk 4]
```diff
@@ -0,0 +1,24 @@
+{% extends "base.html" %}
+
+{% block page_title %}{{ _('Group terms') }}{% endblock %}
```
Comment (akatsoulas) [inline mozillians/templates/groups/terms.html]: s/Group terms/Review Terms ?

**Comment by akatsoulas** [comment]
You are missing a migration here. Please amend the changes in the `0006` migration.

File: mozillians/groups/forms.py [hunk 5]
```diff
@@ -40,7 +41,7 @@ class Meta:
         model = Group
         fields = ['name', 'description', 'irc_channel',
                   'website', 'wiki', 'accepting_new_members',
-                  'new_member_criteria']
+                  'new_member_criteria', 'terms']
```
Comment (akatsoulas) [inline mozillians/groups/forms.py]: Why is this needed here?

File: mozillians/groups/forms.py [hunk 6]
```diff
@@ -72,3 +74,17 @@ def clean_filtr(self):
         if self.cleaned_data['filtr'] == '':
             return 'all'
         return self.cleaned_data['filtr']
+
+
+class GroupMembershipForm(forms.ModelForm):
+    class Meta:
+        model = GroupMembership
+        fields = ['terms_accepted', ]
```
Comment (akatsoulas) [inline mozillians/groups/forms.py]: Please remove the comma

File: mozillians/groups/forms.py (hunk 5 above)
Comment (johngian) [inline mozillians/groups/forms.py]: This is GroupForm. ``git diff`` is misleading

File: mozillians/groups/views.py [hunk 7]
```diff
@@ -265,6 +273,32 @@ def edit(request, url, alias_model, template):
     return render(request, alias_model, template)
 
 
+def group_terms(request, url):
```
Comment (akatsoulas) [inline mozillians/groups/views.py]: I would go for something more descriptive. Like review_group_terms or plain review_terms

File: mozillians/groups/views.py [hunk 8]
```diff
@@ -354,6 +388,10 @@ def group_add_edit(request, url=None):
         # Ensure curator is in the group when it's created
         if profile == group.curator and not group.has_member(profile):
             group.add_member(profile)
+            if group.terms:
```
Comment (akatsoulas) [inline mozillians/groups/views.py]: This is rather interesting. Although it makes sense to auto-accept the terms for a curator, at the same time you are accepting terms that you have never reviewed. Moreover if you factor in PR #1250, the you have curators adding new curators where the new curators have the terms accepted without even logging into the portal. I don't have a clear suggestion here, I am just flagging this.  Thoughts? cc @ppapadeas

**Comment by akatsoulas** [comment]
Should this functionality be available only for groups that require approval to join or every group?

**Comment by akatsoulas** [comment]
Right after a request to join a group, I am redirected to a page to accept/decline the group terms. This page should only be displayed right after the curator has accepted my request.

File: mozillians/groups/views.py (hunk 8 above)
Comment (johngian) [inline mozillians/groups/views.py]: I suggest to treat curators the same way with the rest of the users. And if another UX comes we can ammend this.

Comment (akatsoulas) [inline mozillians/groups/views.py]: Well if this is the case, then we should not auto-accept the terms for the curators as we do here. We should block their curator status until they accept the terms. The only exception should be the creator of the group.

**Comment by johngian** [comment]
* This functionality should be required for every group with terms not only those requiring request.

**Comment by johngian** [comment]
```
Right after a request to join a group, I am redirected to a page to accept/decline the group terms. This page should only be displayed right after the curator has accepted my request.
```

This was the initial implementation but I changed it. It makes sense though to be that way.

File: mozillians/groups/views.py (hunk 8 above)
Comment (ppapadeas) [inline mozillians/groups/views.py]: When adding a curator it is obvious that this is a trusted person (already under multiple NDAs). Please proceed with auto adding to the group and auto accepting the legal text.

**Comment by johngian** [comment]
PR updated:
* Code styling changes
* ``Review terms`` page is only shown to ``GroupMembership.MEMBER``.

File: mozillians/groups/forms.py [hunk 9]
```diff
@@ -72,3 +74,17 @@ def clean_filtr(self):
...
+
+class GroupMembershipForm(forms.ModelForm):
+    class Meta:
+        model = GroupMembership
+        fields = ['terms_accepted']
+        widgets = {
+            'terms_accepted': RadioSelect(
+                choices=[
+                    (True, _('I accept the group terms.')),
+                    (False, _("I don't accept the group terms."))
```
Comment (akatsoulas) [inline mozillians/groups/forms.py]: Maybe something like `I accept these terms` and `I don't accept these terms` is a bit nicer.

File: mozillians/groups/migrations/0007_auto_20151127_0439.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
```
Comment (akatsoulas) [inline mozillians/groups/migrations/0007_auto_20151127_0439.py]: Please squash this migration together with `0006`

File: mozillians/groups/models.py [hunk 11]
```diff
@@ -291,9 +293,12 @@ def remove_member(self, userprofile, send_email=True):
     def has_member(self, userprofile):
         """
         Return True if this user is in this group with status MEMBER.
+        If group has `terms` exclude users that haven't accepted the terms.
         """
-        return self.groupmembership_set.filter(userprofile=userprofile,
-                                               status=GroupMembership.MEMBER).exists()
+        qs = self.groupmembership_set.filter(userprofile=userprofile, status=GroupMembership.MEMBER)
```
Comment (akatsoulas) [inline mozillians/groups/models.py]: I believe this is error prone and will lead to future mistakes/bugs. The membership status should remain `PENDING` until the user has accepted the terms, if exist. The flow should be the following:

* status == MEMBER if there are no terms and users can freely join a group
* status == PENDING if a user can join a group by request and there are no terms
  ** status == MEMBER when the user is accepted
* status == PENDING if a user can join a group by request and there terms
  ** status == MEMBER when the user is accepted *and* has accepted the terms

**Comment by johngian** [comment]
@akatsoulas PR updated
* Replaced ``terms_accepted`` with a new state in ``GroupMembership.status``
* Various minor fixes

File: mozillians/groups/models.py [hunk 12]
```diff
@@ -123,15 +123,17 @@ class GroupMembership(models.Model):
     # Possible membership statuses:
     MEMBER = u'member'
     PENDING = u'pending'  # Has requested to join group, not a member yet
+    TERMS_PENDING = u'terms_pending'
```
Comment (akatsoulas) [inline mozillians/groups/models.py]: Please add a small comment, similar to the line above.

File: mozillians/groups/models.py [hunk 13]
```diff
@@ -260,13 +263,18 @@ def add_member(self, userprofile, status=GroupMembership.MEMBER):
                 # Status changed
                 old_status = membership.status
                 membership.status = status
-                if (old_status, status) == (GroupMembership.PENDING, GroupMembership.MEMBER):
+                statuses = [(GroupMembership.PENDING, GroupMembership.MEMBER),
+                            (GroupMembership.PENDING, GroupMembership.TERMS_PENDING),
+                            (GroupMembership.TERMS_PENDING, GroupMembership.MEMBER)]
```
Comment (akatsoulas) [inline mozillians/groups/models.py]: I would suggest to add a few comments explaining the various combinations of statuses that you have here.

File: mozillians/groups/models.py [hunk 14]
```diff
@@ -260,13 +263,18 @@ def add_member(self, userprofile, status=GroupMembership.MEMBER):
...
+                if (old_status, status) in statuses:
                     # Request accepted
```
Comment (akatsoulas) [inline mozillians/groups/models.py]: Update this comment please. This isn't true any more.

File: mozillians/groups/models.py [hunk 15]
```diff
@@ -260,13 +263,18 @@ def add_member(self, userprofile, status=GroupMembership.MEMBER):
...
-                        # Group is functional area, we want to sent this update to Basket.
-                        update_basket_task.delay(userprofile.id)
-                    email_membership_change.delay(self.pk, userprofile.user.pk, old_status, status)
+                    if membership.status in [GroupMembership.PENDING, GroupMembership.MEMBER]:
+                        if self.functional_area:
+                            # Group is functional area, we want to sent this update to Basket.
+                            update_basket_task.delay(userprofile.id)
+                        email_membership_change.delay(self.pk, userprofile.user.pk,
+                                                      old_status, status)
                 # else? never demote people from full member to requested, that doesn't make sense
```
Comment (akatsoulas) [inline mozillians/groups/models.py]: Probably we need to remove this, since a user can be demoted (eg expiration of terms)

File: mozillians/groups/models.py [hunk 16]
```diff
@@ -260,13 +263,18 @@ def add_member(self, userprofile, status=GroupMembership.MEMBER):
...
+                            (GroupMembership.PENDING, GroupMembership.TERMS_PENDING),
+                            (GroupMembership.TERMS_PENDING, GroupMembership.MEMBER)]
+                if (old_status, status) in statuses:
                     # Request accepted
                     membership.save()
-                    if self.functional_area:
-                        # Group is functional area, we want to sent this update to Basket.
-                        update_basket_task.delay(userprofile.id)
-                    email_membership_change.delay(self.pk, userprofile.user.pk, old_status, status)
+                    if membership.status in [GroupMembership.PENDING, GroupMembership.MEMBER]:
```
Comment (akatsoulas) [inline mozillians/groups/models.py]: What about the other statuses? Shouldn't a user get notified when for example is accepted to a group but hasn't yet accepted the terms?

File: mozillians/groups/tests/test_views/test_misc.py [hunk 17]
```diff
@@ -181,6 +181,31 @@ def test_basic_group_creation_as_non_manager(self):
         eq_(data['website'] + '/', group.website)
         eq_(data['wiki'] + '/', group.wiki)
 
+    def test_basic_group_creation_with_terms(self):
+        # Managers have access to all group parameters when creating a group
```
Comment (akatsoulas) [inline mozillians/groups/tests/test_views/test_misc.py]: Probably this comment is irrelevant

File: mozillians/groups/tests/test_views/test_misc.py [hunk 18]
```diff
@@ -181,6 +181,31 @@ def test_basic_group_creation_as_non_manager(self):
...
+            # 'functional_area' not checked
+        }
+        with self.login(user) as client:
+            response = client.post(url, data=data, follow=False)
+        eq_(302, response.status_code)
+        group = GroupAlias.objects.get(name=data['name']).alias
+        eq_(group.terms, 'Example terms')
+        membership = GroupMembership.objects.get(group=group, userprofile=user.userprofile)
+        ok_(membership.status, GroupMembership.TERMS_PENDING)
+
```
Comment (akatsoulas) [inline mozillians/groups/tests/test_views/test_misc.py]: Please add one more test for a group only joined `by request` and check how the status is changed in each step.

File: mozillians/groups/urls.py [hunk 19]
```diff
@@ -12,6 +12,8 @@
     url('^groups/add/$', 'views.group_add_edit', name='group_add'),
     url('^group/(?P<url>[-\w]+)/edit/$', 'views.group_add_edit', name='group_edit'),
     url('^group/(?P<url>[-\w]+)/delete/$', 'views.group_delete', name='group_delete'),
+    url('^group/(?P<url>[-\w]+)/terms/$', 'views.review_terms',
+        name='review_terms'),
```
Comment (akatsoulas) [inline mozillians/groups/urls.py]: This should fit in one line

File: mozillians/groups/views.py [hunk 20]
```diff
@@ -256,7 +266,10 @@ def confirm_member(request, url, user_pk):
         if membership.status == GroupMembership.MEMBER:
             messages.error(request, _('This user is already a member of this group.'))
         else:
-            group.add_member(profile)
+            status = GroupMembership.MEMBER
+            if group.terms:
```
Comment (akatsoulas) [inline mozillians/groups/views.py]: Probably if you use `eiif else` the flow could be more clear.

File: mozillians/groups/views.py [hunk 21]
```diff
@@ -265,6 +278,33 @@ def edit(request, url, alias_model, template):
...
+        return redirect(reverse('groups:show_group', args=[group.url]))
+
+    membership = get_object_or_404(GroupMembership, group=group,
+                                   userprofile=request.user.userprofile,
+                                   status=GroupMembership.TERMS_PENDING)
+
+    membership_form = TermsReviewForm(request.POST or None)
+    if membership_form.is_valid():
+        terms_accepted = membership_form.cleaned_data['terms_accepted']
+        if terms_accepted == 'True':
```
Comment (akatsoulas) [inline mozillians/groups/views.py]: Can't you combine this to:
```python
`if membership_form.cleaned_data['terms_accepted':
     # add membership here
else:
     # delete membership
```

File: mozillians/groups/forms.py (hunk 5 above)
Comment (akatsoulas) [inline mozillians/groups/forms.py]: Please add a placeholder for this field. Either as a widget attribute or in the html

File: mozillians/templates/groups/add_edit.html [hunk 22]
```diff
@@ -68,6 +68,10 @@
       {{ mozillians_field(form.accepting_new_members) }}
     </fieldset>
 
+    <fieldset>
+      {{ mozillians_field(form.terms) }}
```
Comment (akatsoulas) [inline mozillians/templates/groups/add_edit.html]: This should be below the `form.new_member_criteria`

File: mozillians/templates/groups/add_edit.html [hunk 23]
```diff
@@ -68,6 +68,10 @@
...
+    </fieldset>
+
     <fieldset id="id_new_member_criteria_fieldset">
       {{ mozillians_field(form.new_member_criteria, required=True) }}
```
Comment (akatsoulas) [inline mozillians/templates/groups/add_edit.html]: When this field exists, is not aligned with the `terms` field.

**Comment by akatsoulas** [comment]
You should update the `pending_member_count` in admin.py to include members that haven't yet accepted the terms of a group.

**Comment by akatsoulas** [comment]
Probably, the `get_annotated_groups` method in users/models.py needs a small update too.

File: mozillians/groups/models.py (hunk 16 above)
Comment (johngian) [inline mozillians/groups/models.py]: Opening a new bug to track this, since this is out of the scope of this bug.

File: mozillians/groups/views.py (hunk 20 above)
Comment (johngian) [inline mozillians/groups/views.py]: Group terms and else refer to the same case but with a minor change (``status``). If we use ``elif group.terms`` we are going to duplicate ``add_member`` code in ``elif/else``.

Comment (akatsoulas) [inline mozillians/groups/views.py]: :+1:

File: mozillians/templates/groups/add_edit.html (hunk 23 above)
Comment (johngian) [inline mozillians/templates/groups/add_edit.html]: Opening a new bug for that since this is not a regression. This is already a bug in master.

File: mozillians/groups/views.py (hunk 20 above)
Comment (johngian) [inline mozillians/groups/views.py]: :rabbit:

**Comment by johngian** [comment]
@akatsoulas PR updated

File: mozillians/groups/admin.py [hunk 24]
```diff
@@ -162,15 +162,16 @@ class GroupAdmin(GroupBaseAdmin):
     inlines = [GroupAliasInline]
     list_display = ['name', 'curator', 'functional_area', 'accepting_new_members',
                     'members_can_leave', 'visible', 'total_member_count', 'full_member_count',
-                    'pending_member_count']
+                    'pending_member_count', 'terms_pending_member_count']
```
Comment (akatsoulas) [inline mozillians/groups/admin.py]: s/tems_pending_member_count/pending_terms_member_count ?

File: mozillians/groups/admin.py [hunk 25]
```diff
@@ -199,9 +201,13 @@ def full_member_count(self, obj):
         return obj.groupmembership_set.filter(status=GroupMembership.MEMBER).count()
 
     def pending_member_count(self, obj):
-        """Return number of members in group."""
+        """Return number of pending members in group."""
         return obj.groupmembership_set.filter(status=GroupMembership.PENDING).count()
 
+    def terms_pending_member_count(self, obj):
+        """Return number of members in group who haven't accepted terms yet."""
+        return obj.groupmembership_set.filter(status=GroupMembership.TERMS_PENDING).count()
```
Comment (akatsoulas) [inline mozillians/groups/admin.py]: This probably holds true for this too (PENDING_TERMS). Sorry for not catching this earlier.

**Comment by akatsoulas** [comment]
r+wc :beers:
